File: app/src/main/java/com/mlsdev/mlsdevstore/presentaion/store/StoreFragment.java
package com.mlsdev.mlsdevstore.presentaion.store;

import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.Observable;
import androidx.databinding.ObservableField;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.recyclerview.widget.GridLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.mlsdev.mlsdevstore.R;
import com.mlsdev.mlsdevstore.data.model.item.SearchResult;
import com.mlsdev.mlsdevstore.databinding.FragmentStoreBinding;
import com.mlsdev.mlsdevstore.dependency_injection.Named;
import com.mlsdev.mlsdevstore.dependency_injection.module.ProductsAdapterModule;
import com.mlsdev.mlsdevstore.presentaion.bottom_navigation.MainActivity;
import com.mlsdev.mlsdevstore.presentaion.categories.CategoriesActivity;
import com.mlsdev.mlsdevstore.presentaion.fragment.BaseFragment;

import javax.inject.Inject;

public class StoreFragment extends BaseFragment implements SwipeRefreshLayout.OnRefreshListener {
    private FragmentStoreBinding binding;
    private StoreViewModel viewModel;

    @Inject
    ViewModelProvider.Factory viewModelFactory;

    @Inject
    @Named(ProductsAdapterModule.Type.RandomProducts)
    ProductsAdapter productsAdapter;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_store, container, false);
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(StoreViewModel.class);
        viewModel.searchResult.addOnPropertyChangedCallback(onProductsChangedCallback);
        binding.setViewModel(viewModel);
        binding.refreshLayout.setOnRefreshListener(this);
        initRecyclerView();
        viewModel.getProducts();
        errorInViewHandler.subscribeAllErrorCallbacks(viewModel, true);
        return binding.getRoot();
    }

    @Override
    public void onResume() {
        super.onResume();
        setTitle(R.string.app_name);
    }

    private Observable.OnPropertyChangedCallback onProductsChangedCallback = new Observable.OnPropertyChangedCallback() {
        @Override
        public void onPropertyChanged(Observable observable, int i) {
            if (observable instanceof ObservableField) {
                ObservableField<SearchResult> results = (ObservableField<SearchResult>) observable;
                productsAdapter.setData(results.get());
                binding.rvProducts.notifyDataSetChanged();
            }
        }
    };

    private void initRecyclerView() {
        binding.rvProducts.setAdapter(productsAdapter);
        ((GridLayoutManager) binding.rvProducts.getLayoutManager()).setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                int viewType = productsAdapter.getItemViewType(position);

                if (viewType == RandomProductsAdapter.VIEW_TYPE_FOOTER || viewType == RandomProductsAdapter.VIEW_TYPE_HEADER)
                    return 2;
                else if (viewType == ProductsAdapter.VIEW_TYPE_ITEM)
                    return 1;
                else
                    return -1;
            }
        });

        productsAdapter.setOnClickListeners(
                view -> ((MainActivity) getActivity()).selectTab(R.id.navigation_item_account),
                view -> CategoriesActivity.Companion.launch(getContext()),
                view -> viewModel.loadMoreItemsFromRandomCategory()
        );
    }

    @Override
    public void onRefresh() {
        viewModel.refresh();
    }
}

File: app/src/main/java/com/mlsdev/mlsdevstore/presentaion/store/StoreFragment.kt
package com.mlsdev.mlsdevstore.presentaion.store

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.navigation.fragment.NavHostFragment.findNavController
import androidx.recyclerview.widget.GridLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.mlsdev.mlsdevstore.R
import com.mlsdev.mlsdevstore.databinding.FragmentStoreBinding
import com.mlsdev.mlsdevstore.presentaion.fragment.BaseFragment

class StoreFragment : BaseFragment(), SwipeRefreshLayout.OnRefreshListener {

    lateinit var productsAdapter: RandomProductsAdapter
    lateinit var binding: FragmentStoreBinding
    val viewModel: StoreViewModel by lazy { ViewModelProviders.of(this, viewModelFactory).get(StoreViewModel::class.java) }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_store, container, false)
        binding.viewModel = viewModel
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.refreshLayout.setOnRefreshListener(this)
        initRecyclerView()
        initErrorHandler()
    }

    private fun initErrorHandler() {
        errorInViewHandler.observeNetworkError(this, viewModel.networkErrorLiveData)
        errorInViewHandler.observeCommonError(this, viewModel.commonErrorLiveData)
        errorInViewHandler.observeTechError(this, viewModel.technicalErrorLiveData)
        errorInViewHandler.observeAuthError(this, viewModel.authErrorLiveData)
    }

    private fun initRecyclerView() {
        productsAdapter = RandomProductsAdapter { item ->
            findNavController(this).navigate(StoreFragmentDirections.actionStoreFlowFragmentToProductFragment(item))
        }

        binding.rvProducts.adapter = productsAdapter
        (binding.rvProducts.layoutManager as GridLayoutManager).spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                val viewType = productsAdapter.getItemViewType(position)
                return when (viewType) {
                    VIEW_TYPE_FOOTER -> 2
                    VIEW_TYPE_HEADER -> 2
                    VIEW_TYPE_ITEM -> 1
                    else -> -1
                }
            }
        }

        productsAdapter.setOnClickListeners(
                View.OnClickListener { },
                View.OnClickListener { findNavController(this).navigate(R.id.categories_fragment) },
                View.OnClickListener { viewModel.loadMoreItemsFromRandomCategory() }
        )

        viewModel.searchResultLiveData.observe(this, Observer {
            productsAdapter.setData(it)
            binding.rvProducts.notifyDataSetChanged()
        })

        viewModel.getProducts()
    }

    override fun onRefresh() {
        viewModel.refresh()
    }
}
